File: backend/blue/message.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
'''
@Author: reber
@Mail: reber0ask@qq.com
@Date: 2021-01-27 01:01:53
@LastEditTime : 2021-01-31 15:26:38
'''

import re
from flask import Blueprint
from flask import request
from flask import jsonify
from flask import make_response
from sqlalchemy import func

from sqlmodule import session_maker
from sqlmodule import User
from sqlmodule import Message
from sqlmodule import to_dict

from libs.auth import is_login
from libs.auth import is_admin
from libs.auth import get_username
from libs.auth import get_user_id
from libs.auth import get_user_role
from libs.utils import utc2bj

from setting import rss_sqlite_uri
from setting import logger

message_blueprint = Blueprint('message', __name__)


@message_blueprint.route("/tabs", methods=['POST'])
@is_login
def get_msg_tabs():
    """
    获取当前用户可以看到的消息 tabs
    root(可以看到 schedule、user、new_data等，即可以看到所有新消息)
    user(可以看到 new_data，即新番剧、新文章相关的消息)
    """

    access_token = request.headers.get('access_token')
    user_role = get_user_role(access_token)

    if user_role == "root":
        r_data = {"code": 0, "data": {"tabs": ['system', 'user']}}
        return make_response(jsonify(r_data))
    else:
        r_data = {"code": 0, "data": {"tabs": ['user']}}
        return make_response(jsonify(r_data))

@message_blueprint.route("/newmsg", methods=['POST'])
@is_login
def is_has_new_msg():
    """
    获取未读消息条数
    """

    access_token = request.headers.get('access_token')
    user_role = get_user_role(access_token)
    username = get_username(access_token)

    data = list()
    try:
        with session_maker(rss_sqlite_uri) as db_session:
            if user_role == "root":
                unread_count = db_session.query(func.count(Message.id)).filter(
                    Message.status=="unread").scalar()
            else:
                unread_count = db_session.query(func.count(Message.id)).filter(
                    Message.username==username, Message.msg_type=="user",
                    (~Message.action.contains("/api/")) | (user_role=="root"), Message.status=="unread").scalar()
    except Exception as e:
        logger.error(str(e))
        r_data = {"code": 1}
        return make_response(jsonify(r_data), 500)
    else:
        r_data = {"code": 0, "data": {"newmsg": unread_count}}
        return make_response(jsonify(r_data))

@message_blueprint.route("/status/update", methods=['POST'])
@is_login
def update_msg_status():
    """
    更新消息状态：已读/未读
    """
    data = request.get_json()
    update_id = data.get("update_id", "") # 单个更新的 id
    status = data.get("status", "read")

    msg_type = data.get("msg_type", "") # 全部更新的 msg_type
    update_all = data.get("update_all", "")
    update_id_list = data.get("update_id_list", "") # 更新选中时的 id 列表

    if update_id:
        update_id_list = [update_id]

    access_token = request.headers.get('access_token')
    user_role = get_user_role(access_token)
    username = get_username(access_token)

    affect_num = 0
    try:
        with session_maker(rss_sqlite_uri) as db_session:
            if update_id_list:
                affect_num = db_session.query(Message).filter(
                    (Message.username == username) | (user_role=="root"), Message.msg_type==msg_type, Message.id.in_(update_id_list)).update({"status": status}, synchronize_session=False)
            if update_all:
                affect_num = db_session.query(Message).filter(
                    (Message.username == username) | (user_role=="root"), Message.msg_type==msg_type).update({"status": status}, synchronize_session=False)

            # 获取剩余未读条数
            user_unread_count = db_session.query(func.count(Message.id)).filter(
                (Message.username == username) | (user_role=="root"), 
                (~Message.action.contains("/api/")) | (user_role=="root"), Message.status=="unread").scalar()
            schedule_unread_count = db_session.query(func.count(Message.id)).filter(
                user_role=="root", Message.status=="unread").scalar()

            unread_count = user_unread_count + schedule_unread_count
    except Exception as e:
        logger.error(str(e))
        r_data = {"code": 1, "msg": "操作失败"}
        return make_response(jsonify(r_data), 500)
    else:
        if affect_num:
            r_data = {"code": 0, "msg": status, "unread_count": unread_count}
            return make_response(jsonify(r_data), 200)
        else:
            r_data = {"code": 1, "msg": "操作失败"}
            return make_response(jsonify(r_data), 500)

@message_blueprint.route("/user/list", methods=['POST'])
@is_login
def get_user_action_msg():
    """
    获取用户未读消息
    """
    data = request.get_json()
    pageIndex = data.get("page", 0)
    pageSize = data.get("limit", 0)

    access_token = request.headers.get('access_token')
    user_role = get_user_role(access_token)
    username = get_username(access_token)

    data = list()
    count = 0
    try:
        with session_maker(rss_sqlite_uri) as db_session:
            unread_count = db_session.query(func.count(Message.id)).filter(
                (Message.username == username) | (user_role=="root"), Message.msg_type == "user", 
                (~Message.action.contains("/api/")) | (user_role=="root"), Message.status=="unread").scalar()
            count = db_session.query(func.count(Message.id)).filter(
                (Message.username == username) | (user_role=="root"), Message.msg_type == "user",
                (~Message.action.contains("/api/")) | (user_role=="root")).scalar()
            results = db_session.query(Message).filter(
                (Message.username == username) | (user_role=="root"), Message.msg_type == "user",
                (~Message.action.contains("/api/")) | (user_role=="root")).order_by(Message.id.desc()).limit(pageSize).offset((pageIndex-1)*pageSize).all()
            data = to_dict(results)
        for index,value in enumerate(data):
            curr_data = data[index]["data"]
            curr_data = re.sub(r'\'password\': \'.*?\'', "'password': '******'", curr_data)
            curr_data = re.sub(r'\'email\': \'.*?\'', "'email': '******'", curr_data)
            data[index]["data"] = curr_data
            data[index]["add_time"] = utc2bj(data[index]["add_time"])
    except Exception as e:
        logger.error(str(e))
        r_data = {"code": 1}
        return make_response(jsonify(r_data), 500)
    else:
        r_data = {"code": 0, "data": data, "count": count, "unread_count": unread_count}
        return make_response(jsonify(r_data), 200)

@message_blueprint.route("/system/list", methods=['POST'])
@is_login
@is_admin
def get_system_msg():
    """
    获取系统未读消息
    """
    data = request.get_json()
    pageIndex = data.get("page", 0)
    pageSize = data.get("limit", 0)

    access_token = request.headers.get('access_token')
    user_id = get_user_id(access_token)
    user_role = get_user_role(access_token)

    data = list()
    count = 0
    try:
        with session_maker(rss_sqlite_uri) as db_session:
            unread_count = db_session.query(func.count(Message.id)).filter(
                Message.msg_type == "system", Message.status=="unread").scalar()
            count = db_session.query(func.count(Message.id)).filter(
                Message.msg_type == "system").scalar()
            results = db_session.query(Message).filter_by(
                msg_type="system").order_by(Message.id.desc()).limit(pageSize).offset((pageIndex-1)*pageSize).all()
            data = to_dict(results)
        for index,value in enumerate(data):
            data[index]["add_time"] = utc2bj(data[index]["add_time"])
    except Exception as e:
        logger.error(str(e))
        r_data = {"code": 1}
        return make_response(jsonify(r_data), 500)
    else:
        r_data = {"code": 0, "data": data, "count": count, "unread_count": unread_count}
        return make_response(jsonify(r_data), 200)

@message_blueprint.route("/delete", methods=['POST'])
@is_login
def delete_msg():
    data = request.get_json()
    delete_id = data.get("delete_id", "") # 单个删除的 id

    msg_type = data.get("msg_type", "") # 全部删除的 msg_type
    delete_all = data.get("delete_all", "")
    delete_id_list = data.get("delete_id_list", "") # 删除选中时的 id 列表

    if delete_id:
        delete_id_list = [delete_id]

    access_token = request.headers.get('access_token')
    user_role = get_user_role(access_token)
    username = get_username(access_token)

    affect_num = 0
    try:
        with session_maker(rss_sqlite_uri) as db_session:
            if delete_id_list:
                affect_num = db_session.query(Message).filter(
                    (Message.username == username) | (user_role=="root"), Message.id.in_(delete_id_list)).delete(synchronize_session=False)
            if delete_all:
                affect_num = db_session.query(Message).filter(
                    (Message.username == username) | (user_role=="root"), Message.msg_type==msg_type).delete(synchronize_session=False)

            # 获取剩余未读条数
            user_unread_count = db_session.query(func.count(Message.id)).filter(
                (Message.username == username) | (user_role=="root"), Message.status=="unread").scalar()
            schedule_unread_count = db_session.query(func.count(Message.id)).filter(
                user_role=="root", Message.status=="unread").scalar()

            if user_role == "root":
                unread_count = user_unread_count + schedule_unread_count
            else:
                unread_count = user_unread_count
    except Exception as e:
        logger.error(str(e))
        r_data = {"code": 1, "msg": "delete error"}
        return make_response(jsonify(r_data), 500)
    else:
        if affect_num:
            r_data = {"code": 0, "msg": "delete success", "unread_count": unread_count}
            return make_response(jsonify(r_data), 200)
        else:
            r_data = {"code": 1, "msg": "delete error"}
            return make_response(jsonify(r_data), 500)

File: routers/message.go
/*
 * @Author: reber
 * @Mail: reber0ask@qq.com
 * @Date: 2022-01-04 20:53:07
 * @LastEditTime: 2022-09-20 10:17:06
 */
package routers

import (
	"github.com/gin-gonic/gin"
	"github.com/reber0/go-common/utils"
	"github.com/reber0/rss-sub/global"
	"github.com/reber0/rss-sub/middleware"
	"github.com/reber0/rss-sub/mydb"
	"gorm.io/gorm"
)

// Message 相关路由
func MessageRouter(r *gin.Engine) {
	msgGroup := r.Group("/api/message").Use(middleware.JWTAuth())
	{
		msgGroup.POST("/tabs", msgTabs)
		msgGroup.POST("/count", msgCount)
		msgGroup.POST("/user_list", msgUserList)
		msgGroup.POST("/api_list", msgApiList)
		msgGroup.POST("/update", msgUpdate)
		msgGroup.POST("/read_all", msgReadAll)
		msgGroup.POST("/delete", msgDelete)
		msgGroup.POST("/delete_all", msgDeleteAll)
	}
}

// 显示消息的 tab
func msgTabs(c *gin.Context) {
	userId := c.GetString("uid")
	_, role := GetUserMsg(userId)

	if role == "root" {
		c.JSON(200, gin.H{
			"code": 0,
			"data": gin.H{
				"tabs": []string{"api", "user"},
			},
		})
	} else {
		c.JSON(200, gin.H{
			"code": 0,
			"data": gin.H{
				"tabs": []string{"user"},
			},
		})
	}
}

// 获取未读消息条数
func msgCount(c *gin.Context) {
	userId := c.GetString("uid")
	uname, role := GetUserMsg(userId)

	var count int64
	if role == "root" {
		global.Db.Model(&mydb.Message{}).Where("status='unread'").Count(&count)
	} else if role == "user" {
		global.Db.Model(&mydb.Message{}).Where("status='unread' and uname=? and action like '%更新%'", uname).Count(&count)
	}

	c.JSON(200, gin.H{
		"code":         0,
		"unread_count": count,
	})
}

// 列用户未读消息，主要是更新的消息
func msgUserList(c *gin.Context) {
	type PostData struct {
		PageIndex int `form:"page" json:"page"`
		PageSize  int `form:"limit" json:"limit"`
	}

	type RespData struct {
		ID        int    `json:"id"`
		Uname     string `json:"username" gorm:"column:uname; type:varchar(50); comment:msg_type 为 user 时是用户名，为 system 时是 schedule/error"`
		Action    string `json:"action" gorm:"column:action; type:text; comment:执行的操作触发的 URI、计划任务动作"`
		Data      string `json:"data" gorm:"column:data; type:text; comment:POST 的数据、得到的数据"`
		Status    string `json:"status" gorm:"column:status; type:varchar(10); comment:状态，是否已读、已看"`
		CreatedAt string `json:"created_at" gorm:"column:created_at; comment:添加时间"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "查询失败",
		})
	} else {
		userId := c.GetString("uid")
		uname, role := GetUserMsg(userId)

		var count int64
		var datas []RespData
		result := global.Db.Model(&mydb.Message{}).Where(
			"(uname=? or ?='root') and action like '%更新%'", uname, role).Count(&count).Order("id desc").Count(&count).Limit(postJson.PageSize).Offset((postJson.PageIndex - 1) * postJson.PageSize).Find(&datas)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "查询失败",
			})
			return
		}

		var unread_count int64
		result = global.Db.Model(&mydb.Message{}).Where(
			"(uname=? or ?='root') and action like '%更新%' and status='unread'", uname, role).Count(&unread_count)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "查询失败",
			})
			return
		}

		for index, data := range datas {
			datas[index].CreatedAt = utils.Unix2Str(data.CreatedAt)
		}

		c.JSON(200, gin.H{
			"code":         0,
			"data":         datas,
			"count":        count,
			"unread_count": unread_count,
		})
	}
}

// 列系统相关未读消息，主要是 api 访问记录消息
func msgApiList(c *gin.Context) {
	type PostData struct {
		PageIndex int `form:"page" json:"page"`
		PageSize  int `form:"limit" json:"limit"`
	}

	type RespData struct {
		ID        int    `json:"id"`
		Uname     string `json:"username" gorm:"column:uname; type:varchar(50); comment:msg_type 为 user 时是用户名，为 system 时是 schedule/error"`
		Action    string `json:"action" gorm:"column:action; type:text; comment:执行的操作触发的 URI、计划任务动作"`
		Data      string `json:"data" gorm:"column:data; type:text; comment:POST 的数据、得到的数据"`
		Status    string `json:"status" gorm:"column:status; type:varchar(10); comment:状态，是否已读、已看"`
		CreatedAt string `json:"created_at" gorm:"column:created_at; comment:添加时间"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "查询失败",
		})
	} else {
		userId := c.GetString("uid")
		uname, role := GetUserMsg(userId)

		var count int64
		var datas []RespData
		result := global.Db.Model(&mydb.Message{}).Where(
			"(uname=? or ?='root') and action like '/api/%'", uname, role).Count(&count).Order("id desc").Count(&count).Limit(postJson.PageSize).Offset((postJson.PageIndex - 1) * postJson.PageSize).Find(&datas)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "查询失败",
			})
			return
		}

		var unread_count int64
		result = global.Db.Model(&mydb.Message{}).Where(
			"(uname=? or ?='root') and action like '/api/%' and status='unread'", uname, role).Count(&unread_count)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "查询失败",
			})
			return
		}

		for index, data := range datas {
			datas[index].CreatedAt = utils.Unix2Str(data.CreatedAt)
		}

		c.JSON(200, gin.H{
			"code":         0,
			"data":         datas,
			"count":        count,
			"unread_count": unread_count,
		})
	}
}

// 更新一条或几条消息状态
func msgUpdate(c *gin.Context) {
	type PostData struct {
		UpdateIDList []int  `form:"id_list" json:"id_list"`
		Status       string `form:"status" json:"status"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "更新失败",
		})
	} else {
		userId := c.GetString("uid")
		uname, role := GetUserMsg(userId)

		updateIDList := postJson.UpdateIDList
		status := postJson.Status

		result := global.Db.Model(&mydb.Message{}).Where(
			"(uname=? or ?='root') and id in ?", uname, role, updateIDList).Update("status", status)
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "更新失败",
			})
			return
		}

		c.JSON(200, gin.H{
			"code": 0,
			"msg":  "更新成功",
		})
	}
}

// 更新所有消息状态为已读
func msgReadAll(c *gin.Context) {
	type PostData struct {
		MsgType string `form:"msgtype" json:"msgtype"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "更新失败",
		})
	} else {
		userId := c.GetString("uid")
		uname, role := GetUserMsg(userId)

		msgtype := postJson.MsgType

		var result *gorm.DB
		if msgtype == "api" {
			result = global.Db.Model(&mydb.Message{}).Where(
				"(uname=? or ?='root') and action like '/api/%'", uname, role).Update("status", "read")
		} else if msgtype == "user" {
			result = global.Db.Model(&mydb.Message{}).Where(
				"(uname=? or ?='root') and action like '%更新%'", uname, role).Update("status", "read")
		}
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "更新失败",
			})
			return
		}

		c.JSON(200, gin.H{
			"code": 0,
			"msg":  "更新成功",
		})
	}
}

// 删一条或几条消息
func msgDelete(c *gin.Context) {
	type PostData struct {
		DeleteAll    bool  `form:"delete_all" json:"delete_all"`
		DeleteIDList []int `form:"id_list" json:"id_list"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "删除失败",
		})
	} else {
		userId := c.GetString("uid")
		uname, role := GetUserMsg(userId)

		deleteIDList := postJson.DeleteIDList

		result := global.Db.Where(
			"(uname=? or ?='root') and id in ?", uname, role, deleteIDList).Delete(&mydb.Message{})
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "删除失败",
			})
			return
		}

		c.JSON(200, gin.H{
			"code": 0,
			"msg":  "删除成功",
		})
	}
}

// 删除所有消息
func msgDeleteAll(c *gin.Context) {
	type PostData struct {
		MsgType string `form:"msgtype" json:"msgtype"`
	}

	postJson := PostData{}
	if err := c.BindJSON(&postJson); err != nil {
		global.Log.Error(err.Error())
		c.JSON(400, gin.H{
			"code": 400,
			"msg":  "更新失败",
		})
	} else {
		userId := c.GetString("uid")
		uname, role := GetUserMsg(userId)

		msgtype := postJson.MsgType

		var result *gorm.DB
		if msgtype == "api" {
			result = global.Db.Where(
				"(uname=? or ?='root') and action like '/api/%'", uname, role).Delete(&mydb.Message{})
		} else if msgtype == "user" {
			result = global.Db.Where(
				"(uname=? or ?='root') and action like '%更新%'", uname, role).Delete(&mydb.Message{})
		}
		if result.Error != nil {
			global.Log.Error(result.Error.Error())
			c.JSON(500, gin.H{
				"code": 500,
				"msg":  "更新失败",
			})
			return
		}

		c.JSON(200, gin.H{
			"code": 0,
			"msg":  "更新成功",
		})
	}
}
